### PR TITLE
o/devicestate, daemon, client: add params to api for installing optional snaps and components

### PR DESCRIPTION
**Pull request by andrewphelpsj**

### CODE REVIEW

File: client/systems.go [hunk 1]
```diff
@@ -185,7 +185,13 @@ type SystemDetails struct {
 	// Volumes contains the volumes defined from the gadget snap
 	Volumes map[string]*gadget.Volume `json:"volumes,omitempty"`
 
-	StorageEncryption *StorageEncryption `json:"storage-encryption,omitempty"`
+	StorageEncryption *StorageEncryption  `json:"storage-encryption,omitempty"`
+	AvailableOptional availableForInstall `json:"available-optional,omitempty"`
+}
+
+type availableForInstall struct {
```
Comment (pedronis) [inline client/systems.go:192]: this should be exported

File: client/systems.go [hunk 2]
```diff
@@ -220,7 +226,8 @@ type InstallSystemOptions struct {
 
 	// OnVolumes is the volume description of the volumes that the
 	// given step should operate on.
-	OnVolumes map[string]*gadget.Volume `json:"on-volumes,omitempty"`
+	OnVolumes       map[string]*gadget.Volume `json:"on-volumes,omitempty"`
+	OptionalInstall availableForInstall       `json:"optional-install,omitempty"`
```
Comment (pedronis) [inline client/systems.go:230]: we probably need a way to distinguish empty vs not provided, so maybe needs to be a pointer?

**Review by pedronis — COMMENTED** [review]
couple of comments

File: client/systems.go (hunk 2 above)
Comment (andrewphelpsj) [reply client/systems.go:230]: Is that because we will need to continue installing all optional (and present in the seed) things in the model, in the absence of an `optional-install` field?

File: client/systems.go [hunk 3]
```diff
@@ -186,6 +186,20 @@ type SystemDetails struct {
...
+}
+
+// AvailableForInstall contains information about snaps and components that are
+// optional in the system's model, but are available for installation.
+type AvailableForInstall struct {
+	// Snaps contains the names of optional snaps that are available for installation.
+	Snaps []string `json:"snaps,omitempty"`
+	// Components contains a mapping of snap names to lists of the names of
+	// optional components that are available for installation.
+	Components map[string][]string `json:"components,omitempty"`
```
Comment (pedronis) [inline client/systems.go:202]: in this world we could have a  `All bool json"all,omitempty"` so that we can do optional-install: {"all": true} which wouble be equivalent to omitting but is a bit more explicit to indicate that all available optional snaps/components should be installed?

Comment (andrewphelpsj) [reply client/systems.go:202]: I've added this in 8a433a11858b2fc621279f379d8fc67e79ebd27d.

File: overlord/devicestate/devicestate_install_mode_test.go [hunk 4]
```diff
@@ -2477,22 +2477,37 @@ func (s *installStepSuite) TestDeviceManagerInstallFinishTasksAndChange(c *C) {
...
 	err = tskInstallFinish.Get("system-label", &onLabel)
 	c.Assert(err, IsNil)
 	c.Assert(onLabel, Equals, "1234")
+
 	var onVols map[string]*gadget.Volume
 	err = tskInstallFinish.Get("on-volumes", &onVols)
 	c.Assert(err, IsNil)
 	c.Assert(onVols, DeepEquals, mockOnVolumes)
+
+	var gotOptionalInstall devicestate.OptionalInstall
```
Comment (pedronis) [inline overlord/devicestate/devicestate_install_mode_test.go:2507]: we should also test the scenario where we pass nil

File: overlord/devicestate/devicestate.go [hunk 5]
```diff
@@ -1921,11 +1921,17 @@ func InstallFinish(st *state.State, label string, onVolumes map[string]*gadget.V
 	finishTask := st.NewTask("install-finish", fmt.Sprintf("Finish setup of run system for %q", label))
 	finishTask.Set("system-label", label)
 	finishTask.Set("on-volumes", onVolumes)
+	finishTask.Set("optional-install", optionalInstall)
 	chg.AddTask(finishTask)
 
 	return chg, nil
 }
 
+type OptionalInstall struct {
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:1930]: probably worth a short doc comment, also maybe this should be defined before InstallFinish

File: overlord/devicestate/devicestate.go [hunk 6]
```diff
@@ -1921,11 +1921,17 @@ func InstallFinish(st *state.State, label string, onVolumes map[string]*gadget.V
 	finishTask := st.NewTask("install-finish", fmt.Sprintf("Finish setup of run system for %q", label))
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:1921]: might be worth already adding a TODO in the install-finish task handler about processing optional-install ?

File: client/systems.go [hunk 7]
```diff
@@ -221,6 +235,17 @@ type InstallSystemOptions struct {
 	// OnVolumes is the volume description of the volumes that the
 	// given step should operate on.
 	OnVolumes map[string]*gadget.Volume `json:"on-volumes,omitempty"`
+	// OptionalInstall contains the optional snaps and components that should be
+	// installed on the system. Omitting this field will result in all optional
+	// snaps and components being installed.
```
Comment (alfonsosanchezbeato) [inline client/systems.go:240]: According to SD168 it is the other way around: "If the “snaps” and “components” fields are empty and “all” is false or omitted, no optional snaps or components are installed"

File: daemon/api_systems.go [hunk 8]
```diff
@@ -325,7 +325,21 @@ func postSystemActionInstall(c *Command, systemLabel string, req *systemActionRe
...
+				}
+			} else {
+				optional = &devicestate.OptionalInstall{
+					Snaps:      req.OptionalInstall.Snaps,
+					Components: req.OptionalInstall.Components,
+				}
+			}
+		}
+
+		chg, err := devicestateInstallFinish(st, systemLabel, req.OnVolumes, optional)
```
Comment (alfonsosanchezbeato) [inline daemon/api_systems.go:342]: Is the case of all=true being considered? In that case optional should be set to the available snaps/components.

File: overlord/devicestate/devicestate.go [hunk 9]
```diff
@@ -1929,8 +1929,13 @@ func InstallFinish(st *state.State, label string, onVolumes map[string]*gadget.V
 	return chg, nil
 }
 
+// OptionalInstall is used to define the snapd and components that are optional
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:1932]: typo: /snapd/snaps/

File: overlord/devicestate/devicestate.go (hunk 5 above)
Comment (pedronis) [reply overlord/devicestate/devicestate.go:1934]: what about the suggestion to move it?

**Review by pedronis — APPROVED** [review]
thank you, couple small things

File: daemon/api_systems.go (hunk 8 above)
Comment (andrewphelpsj) [reply daemon/api_systems.go:342]: If `all == true`, then we pass `nil` to `devicestate.InstallFinish`. I was trying to avoid having two ways of conveying to that function that we should install everything, but maybe that is too implicit?

File: client/systems.go (hunk 7 above)
Comment (andrewphelpsj) [reply client/systems.go:240]: There is a distinction between omitting the entire `optional-install` field (which is what the comment is about) and leaving everything inside of `optional-install` empty (which is what that part of the spec is talking about).

File: overlord/devicestate/devicestate.go (hunk 5 above)
Comment (andrewphelpsj) [reply overlord/devicestate/devicestate.go:1934]: Oh sorry, missed that.

File: daemon/api_systems.go (hunk 8 above)
Comment (alfonsosanchezbeato) [reply daemon/api_systems.go:342]: Right, maybe a bit too implicit, but I think that adding a comment would be enough to clarify.

File: client/systems.go (hunk 7 above)
Comment (alfonsosanchezbeato) [reply client/systems.go:240]: Ok, I see - maybe it would be worth it to have this information as part of this comment too to make it easy to remember.

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, thanks for adding the comments
